### ISSUE TITLE
RSpec mocks isn't properly rolling back stubbed any_instance methods with ruby 2.3.0-preview2

### ISSUE THREAD

**Issue by jrafanie** [rspec/rspec-mocks#1042]
The gist below causes randoms errors depending on the order of the tests but shows that with either rspec 2 or 3 ruby 2.3.0-preview2 randomly fails because the super method is wrong in a different test after a prior test stubbed the method with any_instance.

https://gist.github.com/jrafanie/c94e4c8a4f940e71c9e6

The gist passes with ruby 2.2.3 every time I have run it.

I think the problem is somewhere in the recorder but I have no idea what to try next.
https://github.com/rspec/rspec-mocks/blob/master/lib/rspec/mocks/any_instance/recorder.rb

**Comment by JonRowe** [comment]
It's likely theres a bug caused by Ruby 2.3.0, historically this does happen with preview Rubies but is normally sorted by release day. If this still occurs in final we'll fix it naturally :)

**Comment by jrafanie** [comment]
@tenderlove This is the bug with 2.3.0-preview2 I mentioned... maybe you can help me on this one?  :pray: I can't tell if it's a ruby bug yet, hence it's here...

This looks like a possible cause:  https://github.com/ruby/ruby/blob/v2_3_0_preview2/ChangeLog#L5720-L5834

**Comment by jrafanie** [comment]
Seems related to the #1041

**Comment by tenderlove** [comment]
@jrafanie I'm looking in to it! :)

**Comment by jrafanie** [comment]
Thanks :heart:, let me know if I can help.

**Comment by tenderlove** [comment]
I bisected the problem to ruby/ruby@c8854d2ca4be9ee6946e6d17b0e17d9ef130ee81

I'll try to reduce the repro case so we can report an issue. /cc @ko1

**Comment by tenderlove** [comment]
(though it could very well be a bug in rspec-mock, judging from the commit message)

**Comment by myronmarston** [comment]
Thanks for helping with this, @tenderlove!

**Comment by tenderlove** [comment]
I'm not sure how to fix this, but I think the behavior in Ruby 2.3 is the *correct* behavior.

First, `backup_method!` [copies the method we'd like to stub by using an alias](https://github.com/rspec/rspec-mocks/blob/b94e242ebbc51081fc2245e37c574f5d6d02b7f8/lib/rspec/mocks/any_instance/recorder.rb#L206-L211).  Since `@klass` in this case is `SmallCircle` the `area` method is **copied** as `alias_method_name` on to the `SmallCircle` class.

Next rspec-mocks defines it's own version of `area` on the target class (in this case `SmallCircle`).

Finally [the method is restored at some point](https://github.com/rspec/rspec-mocks/blob/b94e242ebbc51081fc2245e37c574f5d6d02b7f8/lib/rspec/mocks/any_instance/recorder.rb#L189-L198) by aliasing the backup method back to the original method name.  But now the child class `SmallCircle` *has it's own copy* of the superclass's method.  Of course since the super class's implementation of `area` doesn't accept a parameter, when `SmallCircle`'s method gets called, it tries to super, and boom, an exception.

I modified @jrafanie's example slightly to highlight the issue [here](https://gist.github.com/tenderlove/183d0d7244f2f2da32aa).  When you run it and it passes, you'll see this in the output:

```
#<UnboundMethod: SmallCircle(Circle)#area>
```

It indicates that the `area` method is implemented on `Circle`.  When it fails, you see this:

```
#<UnboundMethod: SmallCircle#area>
```

showing that the method is now on `SmallCircle`.  If you run with Ruby 2.2.X, you'll see the same behavior of the superclass method moving down to the subclass via [the puts I added](https://gist.github.com/tenderlove/183d0d7244f2f2da32aa#file-boom-rb-L53), but there was essentially a bug in Ruby that caused the `super` call to skip the parent, and the grandparent's method gets called instead.

I think this is a bug in rspec-mocks.  The solution is that the target class for method backup and restore shouldn't be the class that the user passes in, but should be the class where the method is actually defined.  I'm not totally sure how to do that though. :dizzy_face:

**Comment by tenderlove** [comment]
Here is a short script I wrote to demo:

```ruby
class Shape
  def area(value)
    value
  end
end

class Circle < Shape
  def area
    super(2)
  end
end

class SmallCircle < Circle
end

SmallCircle.class_exec do
  alias_method :backup, :area # :area gets copied to `SmallCircle` as :backup
  define_method :area do
    1
  end
end

p SmallCircle.new.area

SmallCircle.class_exec do
  remove_method :area
  # :backup gets copied to `SmallCircle` as :area.  Now SmallCircle has a copy
  # of `area` that calls `super`.  Of course `super` is `Circle`, and `area`
  # in Circle doesn't take a parameter
  alias_method :area, :backup
end

p SmallCircle.new.area
```

On Ruby 2.2 it will print 1 and 2, and on Ruby 2.3 it raises an exception.

**Comment by jrafanie** [comment]
Wow, that makes total sense now @tenderlove. Thanks for looking into it. I wonder if there's similar code in other mocking libraries that relied on the bug that is now fixed in 2.3.0. I'm not quite sure how to fix it. I'll need to think about it.

**Comment by jrafanie** [comment]
@tenderlove This bug fix in 2.3.0-preview2 also breaks minitest-stub_any_instance, see: https://gist.github.com/jrafanie/5f45ce478a024621ff14  cc @codeodor

**Comment by jrafanie** [comment]
The code in https://github.com/codeodor/minitest-stub_any_instance/blob/master/lib/minitest/stub_any_instance.rb looks nearly identical to the way rspec-mocks is doing it as @tenderlove described [above](https://github.com/rspec/rspec-mocks/issues/1042#issuecomment-165619420)

**Comment by jrafanie** [comment]
Ah, it looks like Mocha is unaffected by this change.

https://github.com/freerange/mocha/blob/master/lib/mocha/any_instance_method.rb

I think the problem is we need to back up the method `unless @klass.instance_method(method_name).owner == @klass`.  Right now, we're doing it on the restore_original_method! and it's too late since the method was backed up into `@klass` so it will always be the owner of it.  Working on a PR for everyone to throw :tomato: at.

**Comment by jrafanie** [comment]
Specifically, this is the line in mocha that made me think checking the method owner beforehand is the way to go:  https://github.com/freerange/mocha/blob/599b1dcd3381945d01981d7f99f43aa0e545af12/lib/mocha/any_instance_method.rb#L19

**Comment by myronmarston** [comment]
Thanks for the detailed explanation, @tenderlove, and for looking into this @jrafanie.  These sorts of complexities are why we generally recommend against using any_instance.

@jrafanie -- your idea sounds right.  In this specific example, since `SmallCircle` does not originally have an implementation of `area` defined on its class, I don't think we have to backup and restore the method at all -- we can just allow the new definition to be added, and then remove it, and everything should "just work".

Want to put together a PR?

**Comment by jrafanie** [comment]
Yes, I'm working on it @myronmarston... trying to add a failing test for the existing behavior.

**Comment by tenderlove** [comment]
> I don't think we have to backup and restore the method at all -- we can just allow the new definition to be added, and then remove it, and everything should "just work".

:+1:

It also seems possible that rather than targeting the class the user passed in, use the `owner` on the instance method:

```
irb(main):001:0> class Shape; def area(x); x; end end
=> :area
irb(main):002:0> class Circle < Shape; def area; 2; end end
=> :area
irb(main):003:0> class SmallCircle < Circle; end   
=> nil
irb(main):004:0> SmallCircle.instance_method(:area).owner
=> Circle
irb(main):005:0>
```

Unsure if it will work in all cases though.

**Comment by codeodor** [comment]
@jrafanie thanks for looping me in and providing that demo. I'll work on the implementation there when I get a chance. 

@tenderlove thanks for confirming it's not a Ruby bug.

**Comment by jrafanie** [comment]
@JonRowe @myronmarston @tenderlove please review #1043.

Note, I tried to avoid "the backup/restore" for methods not owned by the class but I think we need to backup this method so we can track when it's invoked, args passed to it, expectations, etc.  It wasn't that obvious how to disable the backup/restore of the method for non-owning classes while still keeping the recorder/invocation and argument tracking happy.

```diff
diff --git a/lib/rspec/mocks/any_instance/recorder.rb b/lib/rspec/mocks/any_instance/recorder.rb
index aa59e3c..5d2b530 100644
--- a/lib/rspec/mocks/any_instance/recorder.rb
+++ b/lib/rspec/mocks/any_instance/recorder.rb
@@ -207,6 +207,12 @@ module RSpec
         end

         def backup_method!(method_name)
+          # TODO: Remove me:  This is what I wanted to do originally,
+          # but most tests fail, maybe because the observer/recorder/playback
+          # expected it???
+          #
+          # return unless @klass.instance_method(method_name).owner == @klass
+
```

The tests completely blew up after I did that.  So, I went with #1043 instead.

**Comment by headius** [comment]
This also works like 2.2 in JRuby because when we alias, we just copy a reference to the original method. The original method still thinks its implementer is Circle and so super'ing proceeds up to Shape rather than doing a "dynamic" super from SmallCircle to Circle. This is an issue with how we handle copying methods around...an issue it seems MRI has fixed in 2.3.

@enebo This is another case where we need to be able to duplicate methods but change implementationClass so that super proceeds from the correct point in the hierarchy, similar to the module.dup issue we've punted on for a while now.
